rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md

```diff
@@ -0,0 +1,542 @@
+    text_vectorizer.adapt(x)
+  
+    # Save the text vectorization model.
+    keras.Sequential([text_vectorizer]).save(
+        os.path.join(self.trial_dir, "text_vectorization_model"))
+  
+    return model.fit(
+        # Convert x from strings to integer vectors.
+        text_vectorizer(x),
+        y,
+        validation_data=validation_data,
+        callbacks=callbacks,
+    )
+```
+	
+
+#### Custom training loop
+
+```py
+class MyHyperModel(kt.HyperModel):
+  def build(self, hp):
+    inputs = keras.Input(shape=(10,))
+    outputs = layers.Dense(hp.Int("units", 16, 128), activation='relu')(inputs)
+    outputs = layers.Dense(1, activation='sigmoid')
+    model = keras.Model(inputs, outputs)
+    return model
+  
+  def fit(self, hp, model, dataset, validation_data, **kwargs):
+    lr = hp.Float("learning_rate", 1e-4, 1e-2, sampling="log", default=1e-3)
+    optimizer = tf.keras.optimizers.Adam(lr)
+    loss_tracker = tf.keras.metrics.Mean()
+    # Track the validation loss
+    val_loss_tracker = tf.keras.metrics.Mean()
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy()
+    # Record the minimum validation loss during fit.
+    min_val_loss = float("inf")
+  
+     @tf.function
+    def run_train_step(data):
+      images = tf.dtypes.cast(data[0], "float32") / 255.0
+      labels = data[1]
+      with tf.GradientTape() as tape:
+        logits = model(images)
+        loss = loss_fn(labels, logits)
+      gradients = tape.gradient(loss, model.trainable_variables)
+      optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+      loss_tracker.update_state(loss)
+  
+     @tf.function
+    def run_val_step(data):
+      images = tf.dtypes.cast(data[0], "float32") / 255.0
+      labels = data[1]
+      logits = model(images)
+      loss = loss_fn(labels, logits)
+      val_loss_tracker.update_state(loss)
+  
+    for epoch in range(2):
+      for batch, data in enumerate(dataset):
+        run_train_step(data)
+      print(f"Epoch loss: {loss_tracker.result().numpy()}")
+      loss_tracker.reset_states()
+      for batch, data in enumerate(validation_data):
+        run_val_step(data)
+      val_loss = val_loss_tracker.result().numpy()
+      min_val_loss = min(min_val_loss, val_loss)
+      print(f"Epoch val_loss: {val_loss}")
+      val_loss_tracker.reset_states()
+  
+    return min_val_loss
+```
+
+You may also subclass `keras.Model` to override `train_step`.
+
+#### Fine tuning with pretrained weights
+
+```py
+class FineTuneModel(keras.Model):
+  def __init__(self, base_model, head):
+    self.base_model = base_model
+    self.head = head
+  def call(self, inputs):
+    return self.head(self.base_model(inputs))
+
+class MyHyperModel(kt.HyperModel):
+
+  def build(self, hp):
+    head = keras.Sequential([
+      layers.GlobalAveragePooling2D(),
+      layers.Dense(hp.Int("units", 32, 128)),
+      layers.Dense(1),
+    ])
+    return FineTuneModel(
+        base_model=keras.applications.ResNet50(
```

Comment:
You can just make the ResNet the first layer in the Sequential (or it could be a functional model as well)